src/Compilers/Core/Portable/InternalUtilities/Debug.cs

```diff
@@ -23,5 +24,43 @@ public static void AssertNotNull<T>([NotNull] T value)
         {
             Assert(value is object, "Unexpected null reference");
         }
+
+        /// <summary>
+        /// Generally <see cref="Debug.Assert(bool)"/> is a sufficient method for enforcing DEBUG 
+        /// only invariants in our code. When it triggers that providse a nice stack trace for 
+        /// investigation. Generally that is enough.
+        /// 
+        /// <para>There are cases for which a stack is not enough and we need a full heap dump to 
+        /// investigate the failure. This method takes care of that. The behavior is that when running
+        /// in our CI environment if the assert triggers we will rudely crash the process and 
+        /// produce a heap dump for investigation.</para>
+        /// </summary>
+        [Conditional("DEBUG")]
+        internal static void AssertOrFailFast([DoesNotReturnIf(false)] bool condition, string? message = null)
+        {
+#if NET20 || NETSTANDARD1_3
+            Debug.Assert(condition);
+#else
+            if (!condition)
+            {
+                if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("HELIX_DUMP_FOLDER")))
```

Comment:
Is this our mechanism for deciding that we're running in a CI environment? I am not sure if this variable is set during the build or during "single-machine" tests.